File: packages/hqtui/src/terminal.ts
import { ansi, moveTo, moveUp, setTitle } from "./ansi.ts";
import { type Capabilities, type CapabilityOverrides, detectCapabilities } from "./capabilities.ts";
import { InputParser, type InputEvent } from "./input.ts";

/**
 * How much of the terminal the app owns.
 *
 * `fullscreen` is what hqtui has always done: the alternate screen, the whole
 * grid, and the user's shell handed back untouched at the end.
 *
 * `inline` draws a bounded strip in the normal flow of the command line, the
 * shape every installer and build tool uses -- a few live rows pinned below
 * output that scrolls away above them, and a readable transcript left behind
 * when the process exits. It has no absolute coordinates it can trust, because
 * it does not know which screen row it started on and the terminal can scroll
 * it up at any moment; it navigates from a saved cursor instead.
 *
 * `fixed` claims a rectangle of a terminal something else is driving.
 */
export type Viewport =
  | { mode: "fullscreen" }
  | { mode: "inline"; height: number }
  | { mode: "fixed"; x: number; y: number; width: number; height: number };

export interface Rect {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface TerminalOptions {
  input?: NodeJS.ReadStream;
  output?: NodeJS.WriteStream;
  /** Use the alternate screen so the user's scrollback survives. Default true. */
  alternateScreen?: boolean;
  mouse?: boolean;
  hideCursor?: boolean;
  bracketedPaste?: boolean;
  focusEvents?: boolean;
  title?: string;
  capabilities?: CapabilityOverrides;
  /** Restore the terminal on SIGINT/SIGTERM/uncaught errors. Default true. */
  installExitHandlers?: boolean;
  /** How long to wait before a lone ESC counts as the Escape key. Default 30ms. */
  escapeTimeout?: number;
  /** How much of the terminal to draw into. Default the whole of it. */
  viewport?: Viewport;
}

export interface TerminalSize {
  columns: number;
  rows: number;
}

type Listener<T> = (value: T) => void;

/**
 * Owns the TTY: raw mode, alternate screen, mouse reporting, and — above all —
 * putting everything back. A crashed app must never leave an unusable shell.
 */
export class Terminal {
  readonly input: NodeJS.ReadStream;
  readonly output: NodeJS.WriteStream;
  readonly capabilities: Capabilities;
  private options: Required<Omit<TerminalOptions, "capabilities" | "title" | "input" | "output" | "escapeTimeout" | "viewport">> & { title?: string };
  private parser = new InputParser();
  private entered = false;
  /** Milliseconds to wait before deciding a lone ESC was the Escape key. */
  readonly escapeTimeout: number;
  private rawWasSet = false;
  private inputListeners = new Set<Listener<InputEvent>>();
  private resizeListeners = new Set<Listener<TerminalSize>>();
  private cleanupHandlers: (() => void)[] = [];
  private teardownListeners = new Set<() => void>();
  private escapeTimer: NodeJS.Timeout | null = null;
  private viewportMode: Viewport = { mode: "fullscreen" };
  /**
   * Rows an inline viewport has reserved below the anchor. Kept because a
   * resize can shrink the terminal under it, and because `restore` has to know
   * how far down to move before handing the shell back.
   */
  private reserved = 0;
  private onData = (chunk: Buffer | string): void => {
    const text = typeof chunk === "string" ? chunk : chunk.toString("utf8");
    this.dispatch(this.parser.parse(text));
    // A lone ESC is only the Escape key once nothing follows it.
    if (this.escapeTimer) clearTimeout(this.escapeTimer);
    this.escapeTimer = null;
    if (this.parser.hasPending) {
      this.escapeTimer = setTimeout(() => {
        this.escapeTimer = null;
        this.dispatch(this.parser.flush());
      }, this.escapeTimeout);
      this.escapeTimer.unref?.();
    }
  };

  private dispatch(events: InputEvent[]): void {
    for (const event of events) {
      for (const listener of this.inputListeners) listener(event);
    }
  }
  private onResize = (): void => {
    const size = this.size();
    for (const listener of this.resizeListeners) listener(size);
  };

  constructor(options: TerminalOptions = {}) {
    this.input = options.input ?? process.stdin;
    this.output = options.output ?? process.stdout;
    this.capabilities = detectCapabilities(options.capabilities ?? {}, process.env, this.output);
    this.escapeTimeout = options.escapeTimeout ?? 30;
    this.viewportMode = options.viewport ?? { mode: "fullscreen" };
    // Only a fullscreen app may take the alternate screen. The whole point of
    // the other two is to leave what is already on the terminal alone.
    const fullscreen = this.viewportMode.mode === "fullscreen";
    this.options = {
      alternateScreen: fullscreen ? options.alternateScreen ?? true : false,
      mouse: options.mouse ?? this.capabilities.mouse,
      hideCursor: options.hideCursor ?? true,
      bracketedPaste: options.bracketedPaste ?? this.capabilities.bracketedPaste,
      focusEvents: options.focusEvents ?? this.capabilities.focusEvents,
      installExitHandlers: options.installExitHandlers ?? true,
      title: options.title,
    };
  }

  size(): TerminalSize {
    // Bun reports 0 for columns and rows on some ptys, and `?? 80` does not
    // catch a zero — which leaves a 0x0 framebuffer that renders nothing at
    // all. Anything not a positive finite number means "ask somewhere else".
    const usable = (value: unknown): number | undefined => {
      const n = Number(value);
      return Number.isFinite(n) && n > 0 ? Math.floor(n) : undefined;
    };
    return {
      columns: usable(this.output.columns) ?? usable(process.env.COLUMNS) ?? 80,
      rows: usable(this.output.rows) ?? usable(process.env.LINES) ?? 24,
    };
  }

  /** How much of the terminal the app owns. */
  get viewport(): Viewport {
    return this.viewportMode;
  }

  /**
   * The region to draw into, in screen cells.
   *
   * For an inline viewport the `y` is a fiction -- it is always 0, because the
   * strip is addressed from its own saved cursor rather than from the top of
   * the screen -- but the width and height are real, and they are what the
   * framebuffer is sized from.
   */
  viewportRect(): Rect {
    const { columns, rows } = this.size();
    const v = this.viewportMode;
    if (v.mode === "inline") {
      // A viewport taller than the terminal would scroll itself off the top
      // every frame, so it gives up the rows it cannot have.
      return { x: 0, y: 0, width: columns, height: Math.max(1, Math.min(v.height, rows)) };
    }
    if (v.mode === "fixed") {
      const x = Math.max(0, Math.min(v.x, Math.max(0, columns - 1)));
      const y = Math.max(0, Math.min(v.y, Math.max(0, rows - 1)));
      return {
        x,
        y,
        width: Math.max(0, Math.min(v.width, columns - x)),
        height: Math.max(0, Math.min(v.height, rows - y)),
      };
    }
    return { x: 0, y: 0, width: columns, height: rows };
  }

  write(data: string): void {
    if (data.length === 0) return;
    this.output.write(data);
  }

  /** Enter full-screen mode. Idempotent. */
  enter(): void {
    if (this.entered) return;
    this.entered = true;

    let setup = "";
    if (this.options.alternateScreen) setup += ansi.alternateScreenOn;
    if (this.options.hideCursor) setup += ansi.cursorHide;
    if (this.options.mouse && this.capabilities.mouse) setup += ansi.mouseOn;
    if (this.options.bracketedPaste) setup += ansi.bracketedPasteOn;
    if (this.options.focusEvents) setup += ansi.focusOn;
    if (this.options.title) setup += setTitle(this.options.title);
    // Only a fullscreen app owns the grid, so only a fullscreen app may wipe
    // it. An inline strip or a fixed region is a guest on somebody else's
    // screen and has no business clearing it.
    if (this.viewportMode.mode === "fullscreen") setup += ansi.clearScreen + ansi.cursorHome;
    this.write(setup);
    if (this.viewportMode.mode === "inline") this.reserveInline();

    if (this.input.isTTY && typeof this.input.setRawMode === "function") {
      this.input.setRawMode(true);
      this.rawWasSet = true;
    }
    this.input.resume?.();
    this.input.setEncoding?.("utf8");
    this.input.on("data", this.onData);
    this.output.on("resize", this.onResize);

    if (this.options.installExitHandlers) this.installExitHandlers();
  }

  /** Put the terminal back exactly as it was found. Safe to call twice. */
  restore(): void {
    if (!this.entered) return;
    this.entered = false;

    if (this.escapeTimer) clearTimeout(this.escapeTimer);
    this.escapeTimer = null;
    this.input.off?.("data", this.onData);
    this.output.off?.("resize", this.onResize);
    if (this.rawWasSet && typeof this.input.setRawMode === "function") {
      this.input.setRawMode(false);
      this.rawWasSet = false;
    }
    this.input.pause?.();

    let teardown = ansi.reset;
    // An inline app leaves its last frame behind as part of the transcript, so
    // the cursor has to come out below it rather than on top of it.
    if (this.viewportMode.mode === "inline" && this.reserved > 0) {
      teardown = toAnchor() + "\r" + "\n".repeat(this.reserved) + teardown;
      this.reserved = 0;
    }
    if (this.options.focusEvents) teardown += ansi.focusOff;
    if (this.options.bracketedPaste) teardown += ansi.bracketedPasteOff;
    if (this.options.mouse) teardown += ansi.mouseOff;
    if (this.options.hideCursor) teardown += ansi.cursorShow;
    teardown += this.options.alternateScreen
      ? ansi.alternateScreenOff
      : this.viewportMode.mode === "inline" ? "" : `\n`;
    this.write(teardown);

    for (const off of this.cleanupHandlers) off();
    this.cleanupHandlers = [];
  }

  /**
   * Make room for an inline viewport and remember where it starts.
   *
   * There is no way to ask where the cursor is without a round trip the caller
   * would have to await, and no way to trust the answer afterwards -- any
   * output scrolls the screen and moves the strip without a word. So the
   * anchor is never a number: it is a saved cursor position, re-saved whenever
   * the strip moves.
   *
   * Printing the newlines first is what reserves the space. If the cursor was
   * near the bottom the terminal scrolls, which is exactly what should happen;
   * walking back up then lands on the strip's first row wherever it ended up.
   */
  private reserveInline(): void {
    const height = this.viewportRect().height;
    this.reserved = height;
    this.write("\r" + "\n".repeat(Math.max(0, height - 1)) + moveUp(height - 1) + "\r");
    this.write(anchor());
  }

  /**
   * Write lines above an inline viewport, permanently.
   *
   * This is the half of inline mode that makes it worth having: finished work
   * scrolls away into the user's scrollback while the live rows stay put. The
   * lines are printed where the strip currently begins and the strip is
   * re-anchored below them, so if that runs off the bottom the terminal scrolls
   * and the oldest lines leave through the top -- into scrollback, which is
   * where they were always going.
   *
   * The caller repaints the viewport afterwards: everything below the anchor is
   * now whatever the terminal happened to shift there.
   */
  insertBefore(lines: string[]): void {
    if (this.viewportMode.mode !== "inline" || lines.length === 0) return;
    const height = this.reserved;
    let out = toAnchor() + "\r";
    for (const line of lines) out += ansi.clearLine + line + ansi.reset + "\r\n";
    // Re-reserve from the new anchor, then walk back to it. Writing the rows
    // is what forces the terminal to scroll if the strip no longer fits, and
    // walking back afterwards finds it wherever the scroll left it.
    out += ansi.clearLine;
    for (let i = 1; i < height; i++) out += "\r\n" + ansi.clearLine;
    out += moveUp(height - 1) + "\r" + anchor();
    this.write(out);
  }

  onInput(listener: Listener<InputEvent>): () => void {
    this.inputListeners.add(listener);
    return () => this.inputListeners.delete(listener);
  }

  onResizeEvent(listener: Listener<TerminalSize>): () => void {
    this.resizeListeners.add(listener);
    return () => this.resizeListeners.delete(listener);
  }

  /**
   * Called when an exit handler tears the terminal down without the owner
   * asking. Whoever is driving a render loop has to know: the alternate screen
   * is gone, so anything it draws next lands in the user's live shell.
   */
  onTeardown(listener: () => void): () => void {
    this.teardownListeners.add(listener);
    return () => this.teardownListeners.delete(listener);
  }

  private installExitHandlers(): void {
    const restore = () => this.restore();

    const onSignal = (signal: NodeJS.Signals) => () => {
      restore();
      process.exit(signal === "SIGINT" ? 130 : 143);
    };
    const onExit = () => restore();
    // Restoring is not negotiable — a crash must not leave an unusable shell —
    // but deciding the process should die is the host's call, not a rendering
    // library's. If the embedding app installed its own handler, hand over once
    // the terminal is safe; only act as the last resort when nobody else will.
    const onError = (event: "uncaughtException" | "unhandledRejection") => (error: unknown) => {
      // Counted before restoring: `restore()` runs the cleanup handlers, and
      // those remove this very listener — so asking afterwards excludes us and
      // one host handler reads as none.
      const handedOver = process.listenerCount(event) > 1;
      restore();
      // Anything still rendering must stop, or it paints into the shell the
      // alternate screen just gave back.
      for (const listener of this.teardownListeners) listener();
      if (handedOver) return;
      // The terminal is usable again, so the stack trace is actually readable.
      console.error(error);
      process.exit(1);
    };
    const onUncaught = onError("uncaughtException");
    const onRejection = onError("unhandledRejection");

    const sigint = onSignal("SIGINT");
    const sigterm = onSignal("SIGTERM");
    const sighup = onSignal("SIGHUP");

    process.on("SIGINT", sigint);
    process.on("SIGTERM", sigterm);
    process.on("SIGHUP", sighup);
    process.on("exit", onExit);
    process.on("uncaughtException", onUncaught);
    process.on("unhandledRejection", onRejection);

    this.cleanupHandlers.push(() => {
      process.off("SIGINT", sigint);
      process.off("SIGTERM", sigterm);
      process.off("SIGHUP", sighup);
      process.off("exit", onExit);
      process.off("uncaughtException", onUncaught);
      process.off("unhandledRejection", onRejection);
    });
  }
}

/**
 * Save the anchor an inline viewport measures from.
 *
 * The pen is reset first so that the saved graphic rendition is always the
 * default one. DECRC restores attributes along with the position, so without
 * that the colour in force at some arbitrary moment would come back with every
 * jump and quietly desynchronise the renderer's model of the terminal.
 */
function anchor(): string {
  return ansi.reset + ansi.cursorSave;
}

/**
 * Go back to the anchor, and immediately save it again.
 *
 * DECRC is a pop rather than a peek in some terminals: restore twice against
 * one save and the second sends the cursor home, which is the top of the
 * user's screen and not remotely where the viewport is. Re-arming after every
 * restore makes the sequence mean the same thing on both kinds.
 */
function toAnchor(): string {
  return ansi.cursorRestore + anchor();
}

export function createTerminal(options: TerminalOptions = {}): Terminal {
  return new Terminal(options);
}

/**
 * Last-resort cleanup for a process that lost its Terminal reference.
 * Safe to call from a signal handler or a REPL after a bad crash.
 */
export function emergencyRestore(output: NodeJS.WriteStream = process.stdout): void {
  output.write(
    ansi.reset + ansi.focusOff + ansi.bracketedPasteOff + ansi.mouseOff +
      ansi.cursorShow + ansi.alternateScreenOff,
  );
  if (process.stdin.isTTY && typeof process.stdin.setRawMode === "function") {
    process.stdin.setRawMode(false);
  }
}

File: ports/python/hqtui/terminal.py
"""Owns the TTY: raw mode, alternate screen, mouse reporting, and — above all —
putting everything back.

A crashed app must never leave an unusable shell.

Python has the easiest job of the four ports here: ``termios``, ``tty``,
``signal``, ``select`` and ``shutil.get_terminal_size`` are all standard
library, correctly implemented per platform. There is no FFI and no subprocess.
"""

from __future__ import annotations

import os
import select
import signal
import sys
from dataclasses import dataclass, field
from typing import NamedTuple

from . import ansi
from .capabilities import Capabilities, CapabilityOverrides, detect_capabilities
from .input import InputParser

__all__ = ["Terminal", "TerminalOptions", "TerminalSize", "emergency_restore"]

try:  # pragma: no cover - the import is the platform test
    import termios
    import tty

    _POSIX = True
except ImportError:  # pragma: no cover - Windows
    termios = None  # type: ignore[assignment]
    tty = None  # type: ignore[assignment]
    _POSIX = False


class TerminalSize(NamedTuple):
    columns: int
    rows: int


@dataclass(slots=True)
class TerminalOptions:
    alternate_screen: bool = True
    """Use the alternate screen so the user's scrollback survives."""
    mouse: bool | None = None
    hide_cursor: bool = True
    bracketed_paste: bool | None = None
    focus_events: bool | None = None
    title: str = ""
    capabilities: CapabilityOverrides = field(default_factory=CapabilityOverrides)
    install_exit_handlers: bool = True
    """Restore the terminal on SIGTERM/SIGHUP and on an uncaught exception."""
    escape_timeout: float = 0.03
    """How long to wait before a lone ESC counts as the Escape key, in seconds."""


_saved_mode: list | None = None
"""The ``termios`` attributes captured when raw mode was entered.

Module-level because the exit handlers have no ``self`` to reach for.
"""

_resized = False
_terminated = 0


def _on_winch(signum, frame) -> None:  # pragma: no cover - signal path
    global _resized
    _resized = True


def _on_terminate(signum, frame) -> None:  # pragma: no cover - signal path
    global _terminated
    _terminated = signum


def _enter_raw_mode() -> bool:
    """cfmakeraw, plus turning off ISIG: Ctrl+C has to arrive as a keystroke,
    because quitting is the application's decision and ``quit_keys`` is
    configurable."""
    global _saved_mode
    if not _POSIX:
        return False
    try:
        fd = sys.stdin.fileno()
        _saved_mode = termios.tcgetattr(fd)
        tty.setraw(fd)
        return True
    except (termios.error, ValueError, OSError):
        return False


def _leave_raw_mode() -> None:
    global _saved_mode
    if not _POSIX or _saved_mode is None:
        return
    try:
        termios.tcsetattr(sys.stdin.fileno(), termios.TCSADRAIN, _saved_mode)
    except (termios.error, ValueError, OSError):  # pragma: no cover
        pass
    _saved_mode = None


def emergency_restore() -> None:
    """Restore the terminal for a process that lost its ``Terminal``.

    Safe to call twice, and safe to call from an exception hook.
    """
    sys.stdout.write(
        ansi.RESET + ansi.FOCUS_OFF + ansi.BRACKETED_PASTE_OFF + ansi.MOUSE_OFF
        + ansi.CURSOR_SHOW + ansi.ALTERNATE_SCREEN_OFF
    )
    sys.stdout.flush()
    _leave_raw_mode()


class Terminal:
    def __init__(self, options: TerminalOptions | None = None) -> None:
        self.options = options or TerminalOptions()
        self.capabilities: Capabilities = detect_capabilities(self.options.capabilities)
        self.escape_timeout = self.options.escape_timeout
        self._parser = InputParser()
        self._entered = False
        self._raw_set = False
        # Bytes that ended mid-character, held until the rest of them arrive.
        self._partial = b""

    @property
    def _mouse(self) -> bool:
        o = self.options.mouse
        return self.capabilities.mouse if o is None else o

    @property
    def _paste(self) -> bool:
        o = self.options.bracketed_paste
        return self.capabilities.bracketed_paste if o is None else o

    @property
    def _focus(self) -> bool:
        o = self.options.focus_events
        return self.capabilities.focus_events if o is None else o

    def size(self) -> TerminalSize:
        """The current window size, falling back the way the reference does: the
        kernel, then ``COLUMNS``/``LINES``, then 80x24.

        Anything that is not a positive number means "ask somewhere else" — some
        ptys report zero, which would otherwise leave a 0x0 framebuffer that
        renders nothing at all.
        """
        try:
            size = os.get_terminal_size(sys.stdout.fileno())
            if size.columns > 0 and size.lines > 0:
                return TerminalSize(size.columns, size.lines)
        except (OSError, ValueError):
            pass

        def from_env(name: str, fallback: int) -> int:
            try:
                v = int(os.environ.get(name, ""))
                return v if v > 0 else fallback
            except ValueError:
                return fallback

        return TerminalSize(from_env("COLUMNS", 80), from_env("LINES", 24))

    def take_resize(self) -> bool:
        """Whether a SIGWINCH has arrived since this was last called."""
        global _resized
        was, _resized = _resized, False
        return was

    def termination_signal(self) -> int:
        """The signal that asked the process to quit, or 0."""
        return _terminated

    def write(self, data: str) -> None:
        if not data:
            return
        sys.stdout.write(data)
        sys.stdout.flush()

    def enter(self) -> None:
        """Enter full-screen mode. Idempotent."""
        if self._entered:
            return
        self._entered = True

        setup = ""
        if self.options.alternate_screen:
            setup += ansi.ALTERNATE_SCREEN_ON
        if self.options.hide_cursor:
            setup += ansi.CURSOR_HIDE
        if self._mouse and self.capabilities.mouse:
            setup += ansi.MOUSE_ON
        if self._paste:
            setup += ansi.BRACKETED_PASTE_ON
        if self._focus:
            setup += ansi.FOCUS_ON
        if self.options.title:
            setup += ansi.set_title(self.options.title)
        setup += ansi.CLEAR_SCREEN + ansi.CURSOR_HOME
        self.write(setup)

        if self.capabilities.tty:
            self._raw_set = _enter_raw_mode()
        if self.options.install_exit_handlers:
            self._install_exit_handlers()

    def restore(self) -> None:
        """Put the terminal back exactly as it was found. Safe to call twice."""
        if not self._entered:
            return
        self._entered = False

        teardown = ansi.RESET
        if self._focus:
            teardown += ansi.FOCUS_OFF
        if self._paste:
            teardown += ansi.BRACKETED_PASTE_OFF
        if self._mouse:
            teardown += ansi.MOUSE_OFF
        if self.options.hide_cursor:
            teardown += ansi.CURSOR_SHOW
        teardown += ansi.ALTERNATE_SCREEN_OFF if self.options.alternate_screen else "\n"
        self.write(teardown)

        if self._raw_set:
            _leave_raw_mode()
            self._raw_set = False

    def poll_input(self, timeout: float = 0.0) -> list:
        """Every input event available within ``timeout`` seconds, decoded.

        A lone ESC is only the Escape key once nothing follows it, so it is held
        back until ``escape_timeout`` has passed with no further bytes. Unlike
        the other ports, this can wait on the file descriptor itself rather than
        on a clock, because ``select`` is right there.
        """
        events: list = []
        received = False

        while True:
            try:
                ready, _, _ = select.select([sys.stdin], [], [], timeout if not received else 0)
            except (OSError, ValueError):  # pragma: no cover - closed stdin
                break
            if not ready:
                break
            try:
                chunk = os.read(sys.stdin.fileno(), 4096)
            except (OSError, ValueError):  # pragma: no cover
                break
            if not chunk:
                break
            received = True
            self._partial += chunk
            # A read can end in the middle of a multi-byte character. Decode
            # what is whole and keep the tail for the next read, or the parser
            # would see a replacement character where a letter belongs.
            text, self._partial = _split_complete_utf8(self._partial)
            events.extend(self._parser.parse(text))

        if not received and self._parser.has_pending:
            events.extend(self._parser.flush())
        return events

    def _install_exit_handlers(self) -> None:
        """Restoring is not negotiable — a crash must not leave an unusable
        shell — but deciding the process should die is the host's call, not a
        rendering library's. So these handlers only record the signal; the app
        loop notices and exits on its own terms.
        """
        if not _POSIX:
            return
        try:
            signal.signal(signal.SIGWINCH, _on_winch)
            signal.signal(signal.SIGTERM, _on_terminate)
            signal.signal(signal.SIGHUP, _on_terminate)
            # SIGINT is delivered only if raw mode failed; with ISIG off, Ctrl+C
            # arrives as a keystroke instead.
            signal.signal(signal.SIGINT, _on_terminate)
        except (ValueError, OSError):  # pragma: no cover - not the main thread
            return

        previous = sys.excepthook

        def hook(exc_type, exc, tb):  # pragma: no cover - exception path
            # The terminal is usable again, so the traceback is readable.
            emergency_restore()
            previous(exc_type, exc, tb)

        sys.excepthook = hook


def _split_complete_utf8(buf: bytes) -> tuple[str, bytes]:
    """Return the longest valid-UTF-8 prefix and whatever is left over, which is
    at most three bytes of a truncated character."""
    end = len(buf)
    # A character is at most four bytes, so at most three can be incomplete.
    for back in range(min(4, len(buf))):
        i = len(buf) - 1 - back
        b = buf[i]
        if b & 0x80 == 0:
            break  # ASCII: nothing pending
        if b & 0xC0 == 0xC0:
            # A lead byte. If the character it starts is not yet complete, hold
            # it back.
            need = 4 if b & 0xF8 == 0xF0 else 3 if b & 0xF0 == 0xE0 else 2
            if len(buf) - i < need:
                end = i
            break
    return buf[:end].decode("utf-8", errors="replace"), buf[end:]
